tests/test_cases/test_cocotb/test_edge_triggers.py

```diff
@@ -0,0 +1,244 @@
+    yield Timer(10)
+    dut.clk <= 0
+    yield Edge(dut.clk)
+    if dut.clk.value.integer != 0:
+        raise TestError("Value should be 0")
+
+
+@cocotb.test()
+def test_fork_and_monitor(dut, period=1000, clocks=6):
+    cocotb.fork(Clock(dut.clk, period).start())
+
+    # Ensure the clock has started
+    yield RisingEdge(dut.clk)
+
+    timer = Timer(period + 10)
+    task = cocotb.fork(count_edges_cycles(dut.clk, clocks))
+    count = 0
+    expect = clocks - 1
+
+    while True:
+        result = yield [timer, task.join()]
+        if count > expect:
+            raise TestFailure("Task didn't complete in expected time")
+        if result is timer:
+            dut._log.info("Count %d: Task still running" % count)
+            count += 1
+        else:
+            break
+    if count != expect:
+        raise TestFailure("Expected to monitor the task %d times but got %d" %
+                          (expect, count))
+    if result != clocks:
+        raise TestFailure("Expected task to return %d but got %s" %
+                          (clocks, repr(result)))
+
+
+@cocotb.coroutine
+def do_clock(dut, limit, period):
+    """Simple clock with a limit"""
+    wait_period = period / 2
+    while limit:
+        yield Timer(wait_period)
+        dut.clk <= 0
+        yield Timer(wait_period)
+        dut.clk <= 1
+        limit -= 1
+
+
+@cocotb.coroutine
+def do_edge_count(dut, signal):
+    """Count the edges"""
+    global edges_seen
+    count = 0
+    while True:
+        yield RisingEdge(signal)
+        edges_seen += 1
+
+
+@cocotb.test()
+def test_edge_count(dut):
+    """Count the number of edges is as expected"""
+    global edges_seen
+    edges_seen = 0
+    clk_period = 100
+    edge_count = 10
+    clock = cocotb.fork(do_clock(dut, edge_count, clk_period))
+    test = cocotb.fork(do_edge_count(dut, dut.clk))
+
+    yield Timer(clk_period * (edge_count + 1))
+
+    if edge_count is not edges_seen:
+        raise TestFailure("Correct edge count failed - saw %d wanted %d" %
+                          (edges_seen, edge_count))
+
+
+@cocotb.test()
+def test_edge_identity(dut):
+    """
+    Test that Edge triggers returns the same object each time
+    """
+
+    re = RisingEdge(dut.clk)
+    fe = FallingEdge(dut.clk)
+    e = Edge(dut.clk)
+
+    assert re is RisingEdge(dut.clk)
+    assert fe is FallingEdge(dut.clk)
+    assert e is Edge(dut.clk)
+
+    # check they are all unique
+    assert len({re, fe, e}) == 3
+    yield Timer(1)
+
+
+@cocotb.test()
+def test_singleton_isinstance(dut):
+    """
+    Test that the result of trigger expression have a predictable type
+    """
+    assert isinstance(RisingEdge(dut.clk), RisingEdge)
+    assert isinstance(FallingEdge(dut.clk), FallingEdge)
+    assert isinstance(Edge(dut.clk), Edge)
+    assert isinstance(NextTimeStep(), NextTimeStep)
+    assert isinstance(ReadOnly(), ReadOnly)
+    assert isinstance(ReadWrite(), ReadWrite)
+
+    yield Timer(1)
```

Comment:
This is maybe a little out of place since it also touches ReadOnly etc, but is better than before, so I don't think is worth worrying about

Reply:
I had thought to split it up, but I was trying to not introduce any new tests/change existing tests in this PR. I'm fine with scope creep on this though, will change.

Reply:
This one is probably more coherent as a single test than it is split up, but I don't care too much.